### PR TITLE
rust: Channel and Schema structs have IDs

### PR DESCRIPTION
**Pull request by james-rms**
#### Breaking changes
- adds `id` members to the crate::Channel and crate::Schema structs.
- refactored and separated `Writer::add_channel` into two methods, `Writer::add_schema` and `Writer::add_channel`.

### Docs
In docstrings, please review if not clear.
### Description

When a message reader recieves schema and channel information, the schema and channel IDs in the MCAP are hidden from them.  This PR adds those so that readers can observe ID information.

Additionally, this PR updates the Writer interface in these ways:

1. The `add_channel(Channel)` method is replaced with `add_channel(schema_id, topic, message_encoding, metadata)`. This was done because the Channel struct now contains a channel ID, and to continue to use it would force the user to come up with their own channel ID before calling this function. 
2. An `add_schema(...)` method is added to register schemas ahead of channels. This is needed to allow the caller to register a new schema without coming up with an ID in advance.
3. The existing `write(Message)` method is kept, but instead of automatically assigning new IDs for the message channel and schema, it uses the IDs passed in. This means that when writing functions that transform MCAP data, the IDs from the input can be kept in the output.

<table><tr><th>Before</th><th>After</th></tr><tr><td>

</td><td>

</td></tr></table>

### CODE REVIEW

File: rust/examples/common/conformance_writer_spec.rs [hunk 1]
```diff
@@ -20,28 +20,26 @@ pub struct Record {
 }
 
 impl Record {
-    pub fn get_field(self: &Self, name: &str) -> &Value {
```
Comment (james-rms) [inline rust/examples/common/conformance_writer_spec.rs:23]: all changes in these files are due to the `clippy --all-targets` change.

File: rust/examples/conformance_writer.rs [hunk 2]
```diff
@@ -38,26 +40,17 @@ fn write_file(spec: &conformance_writer_spec::WriterSpec) {
...
-                    message_encoding: message_encoding.to_string(),
-                    metadata: std::collections::BTreeMap::new(),
-                };
-                writer
-                    .add_channel(&channel)
+                let returned_id = writer
+                    .add_channel(schema_id as u16, topic, message_encoding, &BTreeMap::new())
                     .expect("Couldn't write channel");
-                channels.insert(id, channel);
+                channel_ids.insert(returned_id, id);
```
Comment (james-rms) [inline rust/examples/conformance_writer.rs:46]: this is backwards: need to map from input to output.

File: rust/src/write.rs [hunk 3]
```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
...
+    // its ID is returned.
+    pub fn add_schema(&mut self, name: &str, encoding: &str, data: &[u8]) -> McapResult<u16> {
+        if let Some(&id) = self.schemas.get_by_left(&SchemaContent {
+            name: name.into(),
+            encoding: encoding.into(),
+            data: data.into(),
+        }) {
+            return Ok(id);
         }
+        let next_schema_id = self.schemas.right_values().max().unwrap_or(&0) + 1;
```
Comment (james-rms) [inline rust/src/write.rs:272]: just take len(schemas) + 1

File: rust/src/write.rs [hunk 4]
```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
...
+    ) -> McapResult<u16> {
+        if let Some(&id) = self.channels.get_by_left(&ChannelContent {
+            topic: Cow::Borrowed(topic),
+            schema_id,
+            message_encoding: Cow::Borrowed(message_encoding),
+            metadata: Cow::Borrowed(metadata),
+        }) {
+            return Ok(id);
+        }
+        let next_channel_id = self
```
Comment (james-rms) [inline rust/src/write.rs:330]: just take len(channels) + 1

**Comment by james-rms** [comment]
cc. @mrkline

File: rust/src/write.rs [hunk 5]
```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
...
-        let schema_id = match &chan.schema {
-            Some(s) => self.add_schema(s)?,
-            None => 0,
-        };
-
-        if let Some(id) = self.channels.get(chan) {
-            return Ok(*id);
+    // Adds a schema, returning its ID. If a schema with the same content has been added already,
+    // its ID is returned.
+    pub fn add_schema(&mut self, name: &str, encoding: &str, data: &[u8]) -> McapResult<u16> {
```
Comment (jtbandes) [inline rust/src/write.rs:264]: Thoughts on having a separate struct that's like "ChannelWithoutId"? Pros: avoid having to thread multiple params through different layers, avoid possible confusion between params of the same type.

File: rust/src/write.rs [hunk 6]
```diff
@@ -175,20 +176,35 @@ impl WriteOptions {
...
+struct ChannelContent<'a> {
+    topic: Cow<'a, str>,
+    schema_id: u16,
+    message_encoding: Cow<'a, str>,
+    metadata: Cow<'a, BTreeMap<String, String>>,
+}
+
+#[derive(Hash, PartialEq, Eq)]
+struct SchemaContent<'a> {
+    name: Cow<'a, str>,
```
Comment (jtbandes) [inline rust/src/write.rs:194]: Could you ELI5 why we need Cow here?

**Review by jtbandes — APPROVED** [review]

File: rust/src/write.rs (hunk 6 above)
Comment (james-rms) [reply rust/src/write.rs:194]: i'll add a comment. We use CoW because we want to compare equality between all of these values without neccessarily cloning or moving all of them into a struct to do that.

Comment (jtbandes) [reply rust/src/write.rs:194]: Is there some overhead inherent in the Cow struct itself? Should I think of as being basically a union between a couple different storage options/types? or a pointer to an allocation elsewhere?

Comment (james-rms) [reply rust/src/write.rs:194]: a Cow is a union between a borrowed and owned version of a given type - so for str, it's (String | &str), for bytes it's (&[u8], Vec<u8>). It implements Deref for those types, allowing you to use their comparison, hash implementations against eachother.

  In context, this struct lets me solves this problem:
```
struct left = {
  a: String::new("abc"),
  b: String::new("def")
}

struct right = {
  a: "abc",
  b: "def",
}
if a == b {
   println!("they are the same!")
}
```

File: rust/src/write.rs [hunk 7]
```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
...
+    // its ID is returned.
+    pub fn add_schema(&mut self, name: &str, encoding: &str, data: &[u8]) -> McapResult<u16> {
+        if let Some(&id) = self.schemas.get_by_left(&SchemaContent {
+            name: name.into(),
+            encoding: encoding.into(),
+            data: data.into(),
+        }) {
+            return Ok(id);
         }
+        let next_schema_id = self.schemas.len() as u16 + 1;
```
Comment (gasmith) [inline rust/src/write.rs:272]: Pedantic: We should probably return an error if there are more than 2^16-1 schemas.

File: rust/src/write.rs [hunk 8]
```diff
@@ -233,87 +249,171 @@ impl<'a, W: Write + Seek> Writer<'a, W> {
...
+    ) -> McapResult<u16> {
+        if let Some(&id) = self.channels.get_by_left(&ChannelContent {
+            topic: Cow::Borrowed(topic),
+            schema_id,
+            message_encoding: Cow::Borrowed(message_encoding),
+            metadata: Cow::Borrowed(metadata),
+        }) {
+            return Ok(id);
+        }
+        let next_channel_id = self.channels.len() as u16;
```
Comment (gasmith) [inline rust/src/write.rs:330]: Pedantic: We should probably return an error if there are more than 2^16 channels.

**Review by gasmith — APPROVED** [review]

File: rust/src/write.rs (hunk 5 above)
Comment (james-rms) [reply rust/src/write.rs:271]: The con here is that the struct would need to be defined with explicit generic lifetimes per borrowed argument. In general the borrow checker is more precise around function arguments than it is around struct members, and if such a struct is not defined in a very particular way it may constrain the API in a way we don't expect.

Aside: Iave a personal bias against deploying this pattern too early. IMO a struct should justify its own existence. If it only exists to pack arguments for one particular function, I don't think it's useful.
